src/packages/engine-core/src/NAPIEngine.ts

```diff
@@ -296,67 +299,60 @@ You may have to run ${chalk.greenBright(
     }
   }
   async start(): Promise<void> {
-    await this.setupPromise
-    await this.disconnectPromise
-    if (this.connectPromise) {
-      debug(`already starting: ${this.connected}`)
-      await this.connectPromise
-      if (this.connected) {
+    await this.libraryInstantiationPromise
+    await this.libraryStoppingPromise
+    if (this.libraryStartingPromise) {
+      debug(
+        `library already starting, this.libraryStarted: ${this.libraryStarted}`,
+      )
+      await this.libraryStartingPromise
+      if (this.libraryStarted) {
         return
       }
     }
-    if (!this.connected) {
+    if (!this.libraryStarted) {
       // eslint-disable-next-line no-async-promise-executor
-      this.connectPromise = new Promise(async (res) => {
-        debug('starting')
+      this.libraryStartingPromise = new Promise(async (res) => {
+        debug('library starting')
         await this.engine?.connect({ enableRawQueries: true })
-        this.connected = true
-        debug('started')
+        this.libraryStarted = true
+        debug('library started')
         res()
       })
-      return this.connectPromise
+      return this.libraryStartingPromise
     }
   }
 
+  // TODO Implement hookProcess to trigger stop
   async stop(): Promise<void> {
-    await this.connectPromise
-    await this.currentQuery
-    debug(`stop state ${this.connected}`)
-    if (this.disconnectPromise) {
-      debug('engine is already disconnecting')
-      await this.disconnectPromise
-      if (!this.connected) {
-        this.disconnectPromise = undefined
+    await this.libraryStartingPromise
+    await this.executingQueryPromise
+
+    debug(`library stopping, this.libraryStarted: ${this.libraryStarted}`)
+    if (this.libraryStoppingPromise) {
+      debug('library is already disconnecting')
+      await this.libraryStoppingPromise
+      if (!this.libraryStarted) {
+        this.libraryStoppingPromise = undefined
         return
       }
     }
-    if (this.connected) {
-      // eslint-disable-next-line no-async-promise-executor
-      this.disconnectPromise = new Promise(async (res) => {
-        await this.emitExit()
-        debug('disconnect called')
-        await this.engine?.disconnect()
-        this.connected = false
-        debug('disconnect resolved')
 
-        res()
-      })
-    }
-    return this.disconnectPromise
-  }
-  kill(signal: string): void {
-    if (this.connected && !this.disconnectPromise) {
+    if (this.libraryStarted) {
       // eslint-disable-next-line no-async-promise-executor
-      this.disconnectPromise = new Promise(async (res) => {
-        await this.emitExit()
-        debug(`disconnect called with kill signal ${signal}`)
+      this.libraryStoppingPromise = new Promise(async (res) => {
+        await this.runBeforeStop()
+        await new Promise((r) => setTimeout(r, 5))
```

Comment:
What does this do an why?

Reply:
Seems to prevent the flaky logging tests from failing